feat: Import export environment flags

kyle-ssg opened the pull request:
Thanks for submitting a PR! Please check the boxes below:

- [x] I have run [`pre-commit`](https://docs.flagsmith.com/platform/contributing#pre-commit) to check linting
- [x] I have added information to `docs/` if required so people know about the feature!
- [x] I have filled in the "Changes" section below?
- [x] I have filled in the "How did you test this code" section below?
- [x] I have used a [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) title for this Pull Request

## Changes

- Allows a user to export an environment to a JSON file
- Allows a user to import an export to a project, specifying the merge strategy as either skip or overwrite destructive.
- Both the import and export allow you to preview changes
- Tag filtering now lets you choose between a union and intersection

<img width="1197" alt="image" src="https://github.com/Flagsmith/flagsmith/assets/8608314/95056485-8584-4f79-9059-0cb13336bf80">

<img width="1174" alt="image" src="https://github.com/Flagsmith/flagsmith/assets/8608314/ba6537ae-581a-46d8-b56d-c812efdc5af5">

<img width="849" alt="image" src="https://github.com/Flagsmith/flagsmith/assets/8608314/cc939b7d-2156-4b61-b3eb-ec6880b5587c">

<img width="429" alt="image" src="https://github.com/Flagsmith/flagsmith/assets/8608314/34067eb9-53c3-47be-9b85-b5c5992835b3">


## How did you test this code?

- Exported multiple environments flags
- Imported using both strategies, alternating between new projects and environments

zachaysan commented on frontend/web/components/import-export/FeatureExport.tsx:
```diff
@@ -0,0 +1,215 @@
+  }
+  const timerRef = useRef<NodeJS.Timer>()
+  useEffect(() => {
+    if (
+      exports?.results.find(
+        (featureExport) => featureExport.status === 'PROCESSING',
+      )
+    ) {
+      timerRef.current = setTimeout(() => {
+        refetch()
+      }, 200)
+    }
```
👍🏼

zachaysan commented on frontend/web/components/import-export/FeatureExport.tsx:
```diff
@@ -0,0 +1,215 @@
+            </>
+          }
+        >
+          {
+            'Selecting an environment will determine the feature defaults used when importing to a project'
+          }
```
Hmmm. Maybe this language could be tightened a bit to differentiate the environment people import to vs the other environments that take the defaults from the import.

matthewelwell replied:
Yeah, I just picked up on this. I think @kyle-ssg is maybe confused here. The environment values are used only to overwrite the values in the environment you import to. The default values that exist on the feature in the project of the _exporting_ environment are the ones that are used in other environments in the import project. 

I would suggest that we _either_ change the wording as follows, or just remove this tooltip altogether. 

```suggestion
          {
            'Select the environment to use for the export.'
          }
```

zachaysan commented on frontend/web/components/import-export/FeatureExport.tsx:
```diff
@@ -0,0 +1,215 @@
+            </>
+          }
+        >
+          {
+            'If any tags are selected then the exported file will only include features that are tagged.'
+          }
```
Maybe some clarity that any feature has any of the supplied tags will be included in the export.

zachaysan commented on frontend/web/components/import-export/FeatureImport.tsx:
```diff
@@ -0,0 +1,303 @@
+          const importItem = fileData.find((v) => v.name === projectFlag.name)
+          let newValue = featureState.value
+          let newEnabled = featureState.enabled
+          if (strategy === 'OVERWRITE_DESTRUCTIVE' && importItem) {
+            if (environment === previewEnvironment) {
+              newEnabled = importItem.enabled
+              newValue = importItem.value
+            } else {
+              newEnabled = importItem.default_enabled
+              newValue = importItem.initial_value
+            }
+          }
```
This block needs a comment. Two or three times I thought you implemented it incorrectly, but it's done right. But it will be hard for others to follow the logic in the future.

zachaysan commented on frontend/web/components/import-export/FeatureImport.tsx:
```diff
@@ -0,0 +1,303 @@
+              owners: [],
+              project: ProjectStore.model!.id,
+              tags: [],
+              type: 'STANDARD',
```
It's not always `STANDARD`, right? I think the multivariate option could come in here.

kyle-ssg replied:
Hmm that's a good point, I guess it's 


```
   type: importItem.multivariate?.length
                ? 'MULTIVARIATE'
                : 'STANDARD',
```

zachaysan commented on frontend/web/components/import-export/ImportPage.tsx:
```diff
@@ -0,0 +1,244 @@
+  const getProjectList = (LDKey: string) => {
+    setIsLoading(true)
+    _data
+      .get(`https://app.launchdarkly.com/api/v2/projects`, '', {
+        'Authorization': LDKey,
+      })
```
Here I'm a little confused why LD came into this feature, but it's on the same import page so maybe that is relevant?

kyle-ssg replied:
It hasn't the file was just moved and the git diff here things it's all new code.

zachaysan commented on frontend/web/components/tables/TableTagFilter.tsx:
```diff
@@ -39,18 +45,39 @@ const TableTagFilter: FC<TableFilterType> = ({
+                  value: tagStrategy,
+                }}
+                options={[
+                  {
+                    label: 'Has all',
+                    value: 'INTERSECTION',
+                  },
+                  {
+                    label: 'Has some',
+                    value: 'UNION',
+                  },
```
I'm happy these were added

zachaysan reviewed (approved):
I left a lot of questions and comments, but ultimately the PR looks great overall and I trust your judgement on what to add vs what to ignore from what I surfaced.

novakzaballa reviewed (approved):
It works and the code looks correct

novakzaballa commented on frontend/common/utils/toFormData.ts:
```diff
@@ -0,0 +1,41 @@
+    formData.append(parentKey, data.toString())
+  } else if (typeof data === 'boolean') {
+    formData.append(parentKey, data ? 'true' : 'false')
+  } else if (data === null) {
+    // formData.append(parentKey, 'null')
```
I think the same.

kyle-ssg replied:
No, actually form data doesn't support null - it'd convert it to the string 'null'

kyle-ssg commented on frontend/common/types/requests.ts:
```diff
@@ -130,8 +131,26 @@ export type Req = {
   getSubscriptionMetadata: { id: string }
-  createLaunchDarklyProjectImport: { project_id: string }
-  getLaunchDarklyProjectImport: { project_id: string }
-  getLaunchDarklyProjectsImport: { project_id: string; import_id: string }
+  createLaunchDarklyProjectImport: {
+    project_key: string
+    token: string
+  }
+  getLaunchDarklyProjectImport: { project_id: string; import_id: string }
+  getLaunchDarklyProjectsImport: { project_id: string }
```
This is just a lint fix, nothing new

kyle-ssg commented on frontend/web/components/JSONUpload.tsx:
```diff
@@ -0,0 +1,98 @@
+            <div className='mb-2'>
+              <strong>Select a file or drag and drop here</strong>
+            </div>
+            <div className='text-muted fs-small mb-4'>.json File</div>
```
It was but maybe it's better without

matthewelwell reviewed (commented):
Tested this from a functionality perspective. Seems to work pretty well from what I can tell, but I've highlighted some issues I experienced here: 

<img width="1178" alt="image" src="https://github.com/Flagsmith/flagsmith/assets/14089968/75518369-f135-4649-b057-583caf60ab79">

^ this isn't strictly true when using the 'skip' import strategy. 

The UX as shown in the screencast below is kinda shonky. I'd argue that the small toast message is not enough feedback for the user here - and they can continue hitting the 'Import Project' button as many times as they like, which I imagine 80% of people will do when 'nothing' happens after they click it the first time. 

https://github.com/Flagsmith/flagsmith/assets/14089968/4a54a920-5da7-4608-9933-5e4af24ac2d5

Finally, is there a way to tell if the import is a no-op? When using the skip strategy, I think this is a valid use case. The preview can help a little here but I'd say that it's only really helpful for smaller projects. We may want to consider how we handle projects with very large numbers of features for this functionality. Perhaps a search option or filter by tag?

kyle-ssg commented:
> ^ this isn't strictly true when using the 'skip' import strategy.

Adjusted, I think I can just remove the first part so it's as such: 

<img width="1192" alt="image" src="https://github.com/Flagsmith/flagsmith/assets/8608314/8d6f160e-da53-4e69-bf57-ca2bf68af533">


> The UX as shown in the screencast below is kinda shonky. I'd argue that the small toast message is not enough feedback for the user here - and they can continue hitting the 'Import Project' button as many times as they like, which I imagine 80% of people will do when 'nothing' happens after they click it the first time.

Adjusted, it'll now clear the file and show the following: 

<img width="1191" alt="image" src="https://github.com/Flagsmith/flagsmith/assets/8608314/ea55fe9b-a5c0-41ca-9a6c-585987ec5f9b">

> Finally, is there a way to tell if the import is a no-op? When using the skip strategy, I think this is a valid use case. The preview can help a little here but I'd say that it's only really helpful for smaller projects. We may want to consider how we handle projects with very large numbers of features for this functionality. Perhaps a search option or filter by tag?

Hmm yeah ok, I think I could reuse the compare features page for this and just show changed flags?

<img width="1175" alt="image" src="https://github.com/Flagsmith/flagsmith/assets/8608314/fb0f7f15-6d8d-42e1-bb7c-7992bd76cca3">.

I'm not entirely sure how we'd page these results if we did this since the imported file contains all features. I'm also conscious of the fact we're probably doing none of this for the LD importer?

matthewelwell reviewed (commented):
One more thing that I've noticed, related to larger projects, is that the 'Export Features' / 'Import Features' button being at the bottom is quite painful. I would suggest this button should live above the list of features.

matthewelwell commented:
The feedback is still pretty weak here: 

<img width="1161" alt="image" src="https://github.com/Flagsmith/flagsmith/assets/14089968/31a167bc-b9b7-4f8d-87f7-c3ad9996c2fb">

I would suggest that we add a new API endpoint to get a feature import by it's ID so that the FE can poll it for the status. Currently it gets the entity back on the POST request to create the import but the `"status"` is returned as `"PROCESSING"`. This will eventually change to `"COMPLETE"` I believe, but the FE has no way of knowing this. cc @zachaysan 

This can perhaps be a future improvement, but I do think the feedback at the moment is really weak and, as a user, it would make me confused as to whether anything has happened or not.

kyle-ssg commented on frontend/web/components/import-export/FeatureImport.tsx:
```diff
@@ -0,0 +1,440 @@
+          <div className='my-2'>
+            <strong>Preview</strong>
+          </div>
+          <div className='my-2 col-6 col-lg-4'>
+            <EnvironmentSelect
+              projectId={projectId}
+              onChange={(e) => {
+                setPreviewEnvironment(e)
+              }}
+              value={previewEnvironment}
+            />
+          </div>
```
Im not sure that'd be a good idea for overwrite destructive, since importing does affect other environments

kyle-ssg commented:
@matthewelwell correct it should be showing new flags, it attempts to filter them whilst filtering existing. This should now be fixed
